### PR TITLE
Fermi-Dirac stability improvement

### PR DESCRIPTION
**Pull request by antoine-levitt**

### CODE REVIEW

File: src/Smearing.jl [hunk 1]
```diff
@@ -75,14 +71,32 @@ function occupation_derivative(S::FermiDirac, x)
...
     - (xlogx(f) + xlogx(1 - f))
 end
+function occupation_divided_difference(S::FermiDirac, x, y, εF, temperature)
+    temperature == 0 && return occupation_divided_difference(None(), x, y, εF, temperature)
+    f(z) = occupation(S, (z-εF) / temperature)
+    fder(z) = occupation_derivative(S, (z-εF)/temperature) / temperature
+    # For a stable computation we use
+    # (fx - fy) = fx fy (exp(y) - exp(x)) = fx fy exp(x) expm1(y-x)
+    # which we symmetrize. This can overflow, in which case we fall back to the standard method
+    will_exp_overflow(z1, z2) = abs((z1-z2)/temperature) > log(floatmax(typeof(x))) / 4 # conservative
```
Comment (mfherbst) [inline src/Smearing.jl:91]: You know how I feel about long lines ...

File: src/Smearing.jl [hunk 2]
```diff
@@ -75,14 +71,32 @@ function occupation_derivative(S::FermiDirac, x)
...
 # although that is not especially useful...
+function xlogx(x)
+    iszero(x) ? zero(x) : x * log(x)
+end
 function entropy(S::FermiDirac, x)
     f = occupation(S, x)
     - (xlogx(f) + xlogx(1 - f))
 end
+function occupation_divided_difference(S::FermiDirac, x, y, εF, temperature)
+    temperature == 0 && return occupation_divided_difference(None(), x, y, εF, temperature)
```
Comment (mfherbst) [inline src/Smearing.jl:85]: I think you should abbreviate temperature by `temp` or something like that. That word destroys the flow of reading in this function.

File: src/Smearing.jl (hunk 1 above)
Comment (antoine-levitt) [reply src/Smearing.jl:91]: There's no logical place to break here

File: src/Smearing.jl (hunk 2 above)
Comment (antoine-levitt) [reply src/Smearing.jl]: Good point, done

File: src/Smearing.jl (hunk 1 above)
Comment (mfherbst) [reply src/Smearing.jl]: come on, just assign `floatmax(typeof(x))` to something, or `T = typeof(x)`

**Comment by mfherbst** [comment]
No don't use `T`. We decided to reserve that for element type

File: src/Smearing.jl (hunk 1 above)
Comment (antoine-levitt) [reply src/Smearing.jl]: less of a problem now that temperature is smaller

**Comment by antoine-levitt** [comment]
zero risk of confusion here, and it's just for this module

**Comment by mfherbst** [comment]
We discussed this and decided to drop using `T` for temperature for consistency. We deliberately did it also in this module, remember? I don't think it is a good idea to revert that decision.

**Comment by antoine-levitt** [comment]
I think T is completely fine, but I also can't be bothered fighting this, so temp it is

**Comment by mfherbst** [comment]
Me neither. That's why you fight once and than stick to it :smile:.

**Comment by mfherbst** [comment]
CI takes unusually long ...

**Comment by antoine-levitt** [comment]
Everything's fine locally so it's probably not the PR. Github misbehaving?

**Comment by mfherbst** [comment]
Probably. I'll restart the CI.

**Comment by antoine-levitt** [comment]
It works in the other PR. WTF.

**Pull request merged**